src/mast/mod.rs

```diff
@@ -0,0 +1,1109 @@
+
+        for stmt in stmts {
+            if let Some((stmt, ret_typ)) = self.monomorphize_stmt(typed_fn_env, stmt)? {
+                stmts_mono.push(stmt);
+
+                if ret_typ.is_some() {
+                    return_typ = ret_typ;
+                }
+            }
+        }
+
+        // check the return
+        if let (Some(expected), Some(observed)) = (expected_return, return_typ.clone()) {
+            if !observed.same_as(&expected.kind) {
+                return Err(self.error(
+                    ErrorKind::ReturnTypeMismatch(observed.clone(), expected.kind.clone()),
+                    expected.span,
+                ));
+            }
+        };
+
+        typed_fn_env.pop();
+
+        Ok((stmts_mono, return_typ))
+    }
+
+    /// Monomorphize a statement.
+    pub fn monomorphize_stmt(
+        &mut self,
+        typed_fn_env: &mut MonomorphizedFnEnv,
+        stmt: &Stmt,
+    ) -> Result<Option<(Stmt, Option<TyKind>)>> {
+        let res = match &stmt.kind {
+            StmtKind::Assign { mutable, lhs, rhs } => {
+                let rhs_mono = self.monomorphize_expr(rhs, typed_fn_env)?;
+                let typ = rhs_mono.typ.as_ref().expect("expected a type");
+                let type_info = MTypeInfo::new(typ, lhs.span, None);
+
+                // store the type of lhs in the env
+                typed_fn_env.store_type(&lhs.value, &type_info)?;
+
+                let stmt_mono = Stmt {
+                    kind: StmtKind::Assign {
+                        mutable: *mutable,
+                        lhs: lhs.clone(),
+                        rhs: Box::new(rhs_mono.expr),
+                    },
+                    span: stmt.span,
+                };
+
+                Some((stmt_mono, None))
+            }
+            StmtKind::ForLoop { var, range, body } => {
+                typed_fn_env.store_type(
+                    &var.value,
+                    // because we don't unroll the loop in the monomorphized AST
+                    // no constant value even if it's a BigInt
+                    &MTypeInfo::new(&TyKind::BigInt, var.span, None),
+                )?;
+
+                let start_mono = self.monomorphize_expr(&range.start, typed_fn_env)?;
+                let end_mono = self.monomorphize_expr(&range.end, typed_fn_env)?;
+
+                if start_mono.constant.is_none() || end_mono.constant.is_none() {
+                    return Err(self.error(ErrorKind::InvalidRangeSize, stmt.span));
+                }
+
+                if start_mono.constant.unwrap() > end_mono.constant.unwrap() {
+                    return Err(self.error(ErrorKind::InvalidRangeSize, stmt.span));
+                }
+
+                let range_mono = Range {
+                    start: start_mono.expr,
+                    end: end_mono.expr,
+                    span: range.span,
+                };
+
+                let (stmts_mono, _) = self.monomorphize_block(typed_fn_env, body, None)?;
```

Comment:
I think this is not going to work, if you have a generic function call that's instantiated differently in every iteration,  it will only use the first instantiation on every iteration. For example:

```rust
for i in 2..N {
  do_stuff_on([0; i])
}
```

this will translate to only calling `do_stuff_on([0; 2])`. 

I think the last time we talked about that we said we shouldn't allow generic calls in for loops (as rust doesn't allow that either)

Reply:
indeed, I just tried compiling this and it failed:

```rust
fn create<NN>(arr: [Field; NN]) -> [Field; NN] {
    return [1; NN];
}

fn main() {
    for ii in 1..5 {
        let arr3 = create([0, ii]);
        assert_eq(arr3[0], ii);
    }
}
```

with:

```
    ╭─[generic_for_loop:42:1]
 42 │         let arr3 = create([0, ii]);
 43 │         assert_eq(arr3[0], ii);
    ·         ───────────┬──────────
    ·                    ╰── here
 44 │     }
    ╰────
  help: this assertion failed
```

Reply:
right, this is something we have to check and forbid in TAST

Reply:
it should now forbid generic function call in a for loop